### PR TITLE
luminous: mgr/BaseMgrModule: drop GIL in set_config

### PR DESCRIPTION
**Pull request by xiexingguo**
https://tracker.ceph.com/issues/39424

---

Holding GIL and run mon_command turns out to be easily deadlock-prone.
https://tracker.ceph.com/issues/35985 is a typical report, and here is
another.

Fix by using PyModuleConfig's own lock for protection of concurrent
accesses, which should be sufficient.

Fixes: http://tracker.ceph.com/issues/39040
Signed-off-by: xie xingguo <xie.xingguo@zte.com.cn>
(cherry picked from commit 96b7319cc92e13198e05398921404959005c26d9)

Conflicts:
	- there are new ceph_store_{set,get} commands
          (and many other mgr changes) in Nautilus and later

---------------------------------------------
Still seeing standby MGR deadlock, need:
https://github.com/ceph/ceph/pull/25052/commits/60876ccce4e9de7e138dc04cfac056b29a607b4c
https://github.com/ceph/ceph/pull/25052/commits/03a761c980a1171fe6adf7a463f3ab80ec6d4052
https://github.com/ceph/ceph/pull/25360/commits/61aa7e2e0230bff49e757c32932d1db4bcad5f67

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

**Comment by xiexingguo** [comment]
retest this please

File: src/mgr/ActivePyModules.cc [hunk 1]
```diff
@@ -99,13 +99,13 @@ PyObject *ActivePyModules::get_server_python(const std::string &hostname)
 PyObject *ActivePyModules::list_servers_python()
 {
   PyThreadState *tstate = PyEval_SaveThread();
-  Mutex::Locker l(lock);
-  PyEval_RestoreThread(tstate);
   dout(10) << " >" << dendl;
 
   PyFormatter f(false, true);
```
Comment (xiexingguo) [inline src/mgr/ActivePyModules.cc]: This is wrong, as we can't touch PyFormatter without holding the GIL, which as a result can cause various GC problems.

This has been fixed in https://github.com/ceph/ceph/pull/26445 by Sage.

**Comment by xiexingguo** [comment]
still seeing various MGR deadlocks in our product environment...

@liewegas Minding taking a look? since these backports are non-trivial...

**Comment by xiexingguo** [comment]
Sure.
I've been running these patch series for a couple of days, so far so good.
I'll keep running for a few more days and comment here.

**Comment by yuriw** [comment]
[wip-yuri2-testing-2019-07-15-1956-luminous](https://trello.com/c/5QdUgJyi/667-wip-yuri2-testing-2019-07-15-1956-luminous)

**Pull request merged**